solr/core/src/java/org/apache/solr/update/UpdateLocks.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.update;
+
+import java.io.IOException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import org.apache.lucene.util.BytesRef;
+import org.apache.solr.common.SolrException;
+import org.apache.solr.common.SolrException.ErrorCode;
+import org.apache.solr.util.IOFunction;
+
+/**
+ * Locks associated with updates in connection with the {@link UpdateLog}.
+ *
+ * @lucene.internal
+ */
+public class UpdateLocks {
+  // names are legacy oriented; TODO rename and use EnvUtils
+  private static final String SYS_PROP_BUCKET_VERSION_LOCK_TIMEOUT_MS =
+      "bucketVersionLockTimeoutMs";
+  public static final int DEFAULT_TIMEOUT =
+      Integer.parseInt(System.getProperty(SYS_PROP_BUCKET_VERSION_LOCK_TIMEOUT_MS, "0"));
+
+  private final long docLockTimeoutMs;
+
+  private final ReadWriteLock lock = new ReentrantReadWriteLock(true);
+
+  private final ConcurrentHashMap<BytesRef, LockAndCondition> idToLock = new ConcurrentHashMap<>();
+
+  public UpdateLocks(long docLockTimeoutMs) {
+    this.docLockTimeoutMs = docLockTimeoutMs;
+  }
+
+  public <R> R runWithLock(BytesRef id, IOFunction<Condition, R> function) throws IOException {
+    final var startTimeNanos = System.nanoTime();
+
+    lockForUpdate();
+    try {
+
+      LockAndCondition lock;
+      while (true) {
+        lock = idToLock.computeIfAbsent(id, (k) -> new LockAndCondition());
+        synchronized (lock) {
+          if (lock.refCount < 0) {
```

Comment:
In which conditions could it be < 0?

Reply:
The comment below it says "// is being removed; try again".

Basically when this races with the lock getting removed by another thread that's done with it.  It was in the map but the next instant we see the refCount is too low.

Reply:
Shouldn't it be == 0 ?